test/unit-tests/decorators.ts

```diff
@@ -2,173 +2,437 @@ import * as decoratorsLib from "../../decorators";
+				let errorMessage = "This future throws.";
+
+				$injector.register(moduleName, {
+					propertyName: () => _.map(expectedResults, expectedResult => {
+						return (() => {
+							assert.isFalse(isPostActionExecuted, "Post action MUST NOT be called before all actions are executed.");
+
+							calledActionsCount++;
+							isActionExecuted = calledActionsCount === expectedResults.length;
+							if (calledActionsCount % 2 === 0) {
+								throw new Error(errorMessage);
+							} else {
+								return expectedResult;
+							}
+						}).future<void>()();
+					})
+				});
+
+				let caughtErrors = 0,
+					resolvedPromises = 0;
+
+				// Use new promise that will be resolved when all promises are rejected.
+				// This way we'll be sure all of them are settled and we can verify the postAction is executed.
+				let mainPromise = new Promise(function (onFulfilled: Function, onRejected: Function) {
+					_.each(getPromisesWithPostAction(), (promise: any, index: number) => promise.then((result: any) => {
+						resolvedPromises++;
+						assert.deepEqual(result, expectedResults[index]);
+						if ((caughtErrors + resolvedPromises) === expectedResults.length) {
+							onFulfilled();
+						}
+					}, (err: Error) => {
+						caughtErrors++;
+						assert.deepEqual(err.message, errorMessage, "Error message of rejection should be the specified one.");
+						if ((caughtErrors + resolvedPromises) === expectedResults.length) {
+							onFulfilled();
+						}
+					}));
+				});
+
+				mainPromise
+					.then((result: any) => {
+						assert.isTrue(isPostActionExecuted, "Post action must be executed after all promises are resolved.");
+						assert.isTrue(isActionExecuted, "All actions must be executed after the promise is resolved.");
+						done();
+					})
+					.catch(done);
+			});
+		});
+	});
+
+	describe("exported", () => {
+		let expectedResults: any[] = [
+			"string result",
+			1,
+			{ a: 1, b: "2" },
+			["string 1", "string2"],
+			true,
+			undefined,
+			null
+		];
+
+		_.each(expectedResults, (expectedResult: any) => {
+			it(`returns correct result when function returns ${_.isArray(expectedResult) ? "array" : typeof (expectedResult)}`, () => {
+				$injector.register(moduleName, { propertyName: () => expectedResult });
+				assert.deepEqual($injector.publicApi.__modules__[moduleName], undefined);
+				let promisifiedResultFunction: any = decoratorsLib.exported(moduleName);
+				// Call this line in order to generate publicApi and get the real Promise
+				promisifiedResultFunction({}, propertyName, {});
+				let actualResult: any = $injector.publicApi.__modules__[moduleName][propertyName]();
+				assert.deepEqual(actualResult, expectedResult);
+			});
+
+			it(`passes correct arguments to original function, when argument type is: ${_.isArray(expectedResult) ? "array" : typeof (expectedResult)}`, () => {
+				$injector.register(moduleName, { propertyName: (arg1: string) => arg1 });
```

Comment:
The definition here is not correct, should be `arg1: any`, also could we rename `arg1` to simply `arg`